### PR TITLE
Added multi-hop SONiC upgrade path test case

### PR DESCRIPTION
**Pull request by Ryangwaite**
### Description of PR

Added a new advanced_reboot test case that verifies consecutive warm-upgrades. The upgrade path is specified through the `--multi_hop_upgrade_path` pytest option i.e. `--multi_hop_upgrade_path http://builds.test/sonic-aboot-broadcom-20181130.105.swi,http://builds.test/sonic-aboot-broadcom-20201231.97.swi,http://builds.test/sonic-aboot-broadcom-20230531.22.swi` which will cold reboot into the base image 20181130 then warm-reboot into 20201231 and finally, warm-reboot into 20230531.

To make the test easy to understand, the style, assertions, etc was kept consistent with the existing A->B test in `tests/upgrade_path/test_upgrade_path.py::test_upgrade_path`. Points to note are as follows:
 - There are 3 lifecycle hooks:
   - `base_image_setup` - run only once which is used to cold-boot the device into image A (20181130.105 from example prior)
   - `pre_hop_setup` - run before each hop which is used to prepare the device for warm-boot into image B, C, etc
   - `post_hop_teardown` - run after each hop which can be used for any cleanup or assertions before the next hop starts.
  - There is a separate `ReloadTest` for each warm-reboot - it doesn't span multiple warm-reboots
  - There is a separate log analyzer per warm-reboot with the logs being captured separately for each.

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405

#### What is the motivation for this PR?
Our existing testing can only do a single hop therefore, to test an upgrade path like A-(warm-boot)->B-(warm-boot)->C, we'd do (cold-boot)->A->(warm-boot)->B, then (cold-boot)->B-(warm-boot)->C. This doesn't cover bugs spanning multiple upgrade hops which we've seen in production. This new test addresses the gap.

#### How did you do it?
Took the existing A->B upgrade_tests, duplicated and modified it to support multi-hop upgrades.

#### How did you verify/test it?
Tested in our lab on Arista-7260CX3-D108C8 going through hops 20181130, 20201231 and 20230531
#### Supported testbed topology if it's a new test case?
t0*
### Documentation

Not yet

### CODE REVIEW

**Comment by vaibhavhd** [comment]
@andywongarista , @davidm-arista - can you please help to test this multi hop warmupgrade test in your lab for 202311 branch qualification?

The test has passed in our lab testing, but I wanted to ensure that the public version of this case is complete and works for community before we  merge.

**Comment by andywongarista** [comment]
@Ryangwaite Can you resolve the conflicts? This PR in its current state is difficult for us to apply internally for testing.

**Comment by vaibhavhd** [comment]
@andywongarista are you testing on 202311 branch? As per our discussion from yesterday I think we decided that (for this one time) you will use sonic-mgmt master branch of this testcase to validate 202311 OS version.
So no cherry picking is needed at the moment.

**Comment by andywongarista** [comment]
This PR is conflicting on the master branch as well, due to https://github.com/sonic-net/sonic-mgmt/pull/14672. These are the conflicts I am referring to.

**Comment by Ryangwaite** [comment]
@andywongarista the conflicts have been resolved. You should be able to test it now.

**Comment by vaibhavhd** [comment]
@andywongarista , please update if you have tested the latest iteration on your testbed. Did the test pass?

**Review by ryanzhu706 — APPROVED** [review]

**Comment by andywongarista** [comment]
> @andywongarista , please update if you have tested the latest iteration on your testbed. Did the test pass?

We are running into an issue with the 20181130 -> 20201231 hop (also on 7260). Please check issue 265 on the shared repo for the logs.

File: tests/upgrade_path/utilities.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import logging
+import re
+from tests.common.errors import RunAnsibleModuleFail
+from tests.common.helpers.upgrade_helpers import install_sonic, reboot, check_sonic_version
+
+logger = logging.getLogger(__name__)
+
+
+def set_base_image_a(duthost, localhost, base_image, tbinfo):
```
Comment (saiarcot895) [inline tests/upgrade_path/utilities.py:9]: Can this be renamed to something more descriptive than `set_base_image_a`?

**Comment by andywongarista** [comment]
We were able to verify the test works with hops: 202305->202311->202405

**Review by saiarcot895 — APPROVED** [review]

**Pull request merged**